### PR TITLE
Recursively run StringUtil::decodeEntities on arrays

### PR DESCRIPTION
**Pull request by aschempp**
I got this error in Sentry. Somehow, my version data contains an array in values (I'm using m-vo/group-widget).

Bildschirmfoto 2024-07-26 um 17 39 05

### CODE REVIEW

File: core-bundle/src/Resources/contao/library/Contao/StringUtil.php [hunk 1]
```diff
@@ -221,6 +221,11 @@ public static function substrHtml($strString, $intNumberOfChars)
 	 */
 	public static function decodeEntities($strString, $strQuoteStyle=ENT_QUOTES, $strCharset=null)
 	{
+		if (\is_array($strString))
+		{
+			return array_map(__METHOD__, $strString);
```
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/StringUtil.php]: `$strQuoteStyle` and `$strCharset` needs to be passed as well.

**Review by ausi — CHANGES_REQUESTED** [review]

**Comment by Toflar** [comment]
I think this PR is completely wrong. The method expects a string - you should fix the caller, not the method.

**Review by Toflar — CHANGES_REQUESTED** [review]
Forgot to request changes to prevent merging.

**Comment by aschempp** [comment]
> I think this PR is completely wrong. The method expects a string - you should fix the caller, not the method.

Any suggestion how to do that efficiently? Since it could be recursive on multiple levels?

**Comment by Toflar** [comment]
Why can it be recursive on multiple levels? The method signature is a string as an argument. There is no recursion.

**Comment by ausi** [comment]
> Any suggestion how to do that efficiently? Since it could be recursive on multiple levels?

```php
array_walk_recursive($to[$k], function (&$value) {
	$value = StringUtil::decodeEntities($value);
});
```

**Comment by aschempp** [comment]
Changed in 69c99dfaa31dc881835abf7341064e0e94cc72b3 but I think this is way worse 🤷

**Review by Toflar — APPROVED** [review]
Your previous solution 

- introduced array support on a method that expected a `$strString` input
- introduced array support on a class that's named `StringUtil`

It couldn't have been more wrong than that. I'm fine with the current solution - could've made it re-usable by combining it with `ArrayUtil` for example but imho that's an edge case anyway.

**Comment by fritzmg** [comment]
What is the stack trace of the original error?

**Comment by aschempp** [comment]
> What is the stack trace of the original error?

see the image in the original comment

File: core-bundle/src/Resources/contao/classes/Versions.php [hunk 2]
```diff
@@ -568,8 +568,25 @@ public function compare($blnReturnBuffer=false)
...
+
+							if (\is_array($from[$k])) {
+								array_walk_recursive($from[$k], static function (&$value) {
+									$value = StringUtil::decodeEntities($value);
+								});
+							}
+							else
+							{
+								$from[$k] = StringUtil::decodeEntities($from[$k]);
+							}
```
Comment (ausi) [inline core-bundle/src/Resources/contao/classes/Versions.php:589]:
```suggestion
							$to[$k] = \is_string($to[$k]) ? StringUtil::decodeEntities($to[$k]) : $to[$k];
							$from[$k] = \is_string($from[$k]) ? StringUtil::decodeEntities($from[$k]) : $from[$k];
```

Wouldn’t this be a more appropriate fix as we cannot reproduce where the array is coming from?

Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/Versions.php:589]: Not sure what you mean. It feels perfectly valid to be an array value in a diff field. As far as I know, the diff field also supports that. And it should also decode entities on these values?

Comment (ausi) [reply core-bundle/src/Resources/contao/classes/Versions.php:589]: I probably misunderstood something then. 
How can I reproduce your case?

Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/Versions.php:589]: You need to get array data into `tl_version.data`. I guess it is easiest if you create a new version record and then manipulate `data` with something like `serialize(['field' => ['foo' => 'bar']])`

Comment (fritzmg) [reply core-bundle/src/Resources/contao/classes/Versions.php:589]: I think we'll need an actual reproduction. If I have the time I'll try to reproduce this with a dataset that uses the `m-vo/group-widget`.

Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/Versions.php:589]: @fritzmg did you find some time by accident? 😅

Comment (fritzmg) [reply core-bundle/src/Resources/contao/classes/Versions.php:589]: I tried with an example like [this](https://github.com/m-vo/contao-group-widget?tab=readme-ov-file#tldr) where I then saved the content element with various changes multiple times and then restored to a previous versions. But I was not able to reproduce the issue.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/classes/Versions.php:589]: As discussed in the Contao call, we cannot merge this without a reproduction case.

Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/Versions.php:589]: Closing then since I don't really care

**Comment by fritzmg** [comment]
For what it's worth, I think I do have a stack trace now:

```
ErrorException:
Warning: Array to string conversion

  at vendor\contao\core-bundle\src\Resources\contao\library\Contao\StringUtil.php:224
  at Contao\StringUtil::decodeEntities()
     (vendor\contao\core-bundle\src\Resources\contao\classes\Versions.php:571)
  at Contao\Versions->compare()
     (vendor\isotope\isotope-core\system\modules\isotope\drivers\DC_ProductData.php:426)
  at DC_ProductData->edit()
     (vendor\contao\core-bundle\src\Resources\contao\classes\Backend.php:667)
  at Contao\Backend->getBackendModule()
     (vendor\contao\core-bundle\src\Resources\contao\controllers\BackendMain.php:168)
  at Contao\BackendMain->run()
     (vendor\contao\core-bundle\src\Controller\BackendController.php:49)
  at Contao\CoreBundle\Controller\BackendController->mainAction()
     (vendor\symfony\http-kernel\HttpKernel.php:163)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw()
     (vendor\symfony\http-kernel\HttpKernel.php:75)
  at Symfony\Component\HttpKernel\HttpKernel->handle()
     (vendor\symfony\http-kernel\Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle()
     (public\index.php:44)
```

In this case it's an Isotope product with an `mvo/contao-group-widget` in it.

**Comment by leofeyer** [comment]
Yes, we should have merged the first commit → https://github.com/contao/contao/pull/8190#issuecomment-2747275615

**Pull request closed without merging**